File: src/main/action-recorder.ts
/**
 * User preferences handler
 */

import * as fs from 'fs';
import * as path from 'path';

export interface MouseAction {
  x: number;
  y: number;
  time: number;
  type: 'move' | 'click' | 'doubleclick';
  button?: 'left' | 'right' | 'middle';
}

export class ActionRecorder {
  private recording: MouseAction[] = [];
  private startTime: number = 0;
  private isRecording: boolean = false;
  private clickListeners: Array<() => void> = [];
  private recordingPath: string;

  constructor() {
    const rootDir = process.cwd();
    this.recordingPath = path.join(rootDir, 'user-prefs.json');
  }

  /**
   * Start capturing input
   */
  startRecording(): void {
    if (this.isRecording) {
      return;
    }

    this.recording = [];
    this.startTime = Date.now();
    this.isRecording = true;

    // Trigger IPC handler to start click monitoring
    try {
      const { ipcMain } = require('electron');
      ipcMain.emit('start-click-monitoring');
    } catch (e) {
      // Ignore
    }
    
    this.setupClickDetection();
  }

  /**
   * Stop capturing and save preferences
   */
  stopRecording(): boolean {
    if (!this.isRecording) {
      return false;
    }

    this.isRecording = false;

    try {
      const { ipcMain } = require('electron');
      ipcMain.emit('stop-click-monitoring');
    } catch (e) {
      // Ignore
    }

    this.cleanupClickDetection();

    // Only clicks are recorded, no optimization needed
    // Filter out any move actions from old recordings for consistency
    this.recording = this.recording.filter(action => action.type === 'click' || action.type === 'doubleclick');

    try {
      fs.writeFileSync(this.recordingPath, JSON.stringify(this.recording, null, 2), 'utf-8');
      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Check if preferences exist
   */
  hasRecording(): boolean {
    return fs.existsSync(this.recordingPath);
  }

  /**
   * Load preferences from file
   */
  loadRecording(): MouseAction[] | null {
    try {
      if (!this.hasRecording()) {
        return null;
      }

      const data = fs.readFileSync(this.recordingPath, 'utf-8');
      const actions = JSON.parse(data) as MouseAction[];
      return actions;
    } catch (error) {
      return null;
    }
  }

  /**
   * Delete preferences file
   */
  deleteRecording(): void {
    try {
      if (fs.existsSync(this.recordingPath)) {
        fs.unlinkSync(this.recordingPath);
      }
    } catch (error) {
      // Ignore
    }
  }

  /**
   * Get current status
   */
  getStatus(): { isRecording: boolean; actionCount: number } {
    return {
      isRecording: this.isRecording,
      actionCount: this.recording.length
    };
  }

  /**
   * Set up click detection
   */
  private setupClickDetection(): void {
    try {
      const { ipcRenderer } = require('electron');
      
      ipcRenderer.on('global-mouse-click', (_event: any, data: any) => {
        if (!this.isRecording) {
          return;
        }

        const button = data.button === 1 || data.button === 0 ? 'left' :
                      data.button === 2 ? 'right' :
                      data.button === 3 || data.button === 4 ? 'middle' : 'left';

        this.recordClick(data.x, data.y, button, data.type || 'click');
      });

      this.clickListeners.push(() => {
        try {
          ipcRenderer.removeAllListeners('global-mouse-click');
        } catch (e) {
          // Ignore
        }
      });
    } catch (e) {
      // Main process - clicks handled directly
    }
  }

  /**
   * Capture click event
   */
  recordClick(x: number, y: number, button: 'left' | 'right' | 'middle' = 'left', type: 'click' | 'doubleclick' = 'click'): void {
    if (!this.isRecording) {
      return;
    }

    const clickAction: MouseAction = {
      x,
      y,
      time: Date.now() - this.startTime,
      type,
      button
    };

    this.recording.push(clickAction);
  }

  /**
   * Clean up click detection listeners
   */
  private cleanupClickDetection(): void {
    this.clickListeners.forEach(cleanup => cleanup());
    this.clickListeners = [];
  }

}

File: src/action_recorder.py
"""
User preferences handler - Records mouse clicks for replay
"""

import json
from pathlib import Path
from typing import Optional, List, Dict, Literal
from pynput import mouse


MouseAction = Dict[str, any]  # type: ignore


class ActionRecorder:
    """Records mouse clicks for replay"""

    def __init__(self):
        self.recording: List[MouseAction] = []
        self.start_time: float = 0.0
        self._is_recording: bool = False
        self.listener: Optional[mouse.Listener] = None
        self.recording_path = Path.cwd() / 'user-prefs.json'

    @property
    def is_recording(self) -> bool:
        """Check if currently recording"""
        return self._is_recording

    @is_recording.setter
    def is_recording(self, value: bool) -> None:
        """Set recording state"""
        self._is_recording = value

    def start_recording(self) -> None:
        """Start capturing input"""
        if self.is_recording:
            return

        self.recording = []
        import time
        self.start_time = time.time()
        self._is_recording = True

        def on_click(x: float, y: float, button: mouse.Button, pressed: bool) -> None:
            if not self.is_recording or not pressed:
                return

            button_str = 'left' if button == mouse.Button.left else \
                        'right' if button == mouse.Button.right else 'middle'

            self.record_click(int(x), int(y), button_str, 'click')

        self.listener = mouse.Listener(on_click=on_click)
        self.listener.start()

    def stop_recording(self) -> bool:
        """Stop capturing and save preferences"""
        if not self.is_recording:
            return False

        self._is_recording = False

        if self.listener:
            self.listener.stop()
            self.listener = None

        # Only clicks are recorded
        self.recording = [action for action in self.recording if action['type'] in ('click', 'doubleclick')]

        try:
            with open(self.recording_path, 'w', encoding='utf-8') as f:
                json.dump(self.recording, f, indent=2)
            return True
        except Exception:
            return False

    def has_recording(self) -> bool:
        """Check if preferences exist"""
        return self.recording_path.exists()

    def load_recording(self) -> Optional[List[MouseAction]]:
        """Load preferences from file"""
        try:
            if not self.has_recording():
                return None

            with open(self.recording_path, 'r', encoding='utf-8') as f:
                actions = json.load(f)
            return actions
        except Exception:
            return None

    def delete_recording(self) -> None:
        """Delete preferences file"""
        try:
            if self.recording_path.exists():
                self.recording_path.unlink()
        except Exception:
            pass

    def get_status(self) -> Dict[str, any]:
        """Get current status"""
        return {
            'isRecording': self.is_recording,
            'actionCount': len(self.recording)
        }

    def record_click(self, x: int, y: int, button: Literal['left', 'right', 'middle'] = 'left',
                    action_type: Literal['click', 'doubleclick'] = 'click') -> None:
        """Capture click event"""
        if not self.is_recording:
            return

        import time
        click_action: MouseAction = {
            'x': x,
            'y': y,
            'time': int((time.time() - self.start_time) * 1000),
            'type': action_type,
            'button': button
        }

        self.recording.append(click_action)
